### PR TITLE
i/b/microceph-support: add microceph-support interface

### PR DESCRIPTION
**Pull request by sabaini**
Add separate microceph-support interface in order to allow more types of block devices to be added as an OSD (bcache, indiv. partitions, device mapper)

Also allow access to the rbd sysfs controls for managing rbd-backed block devices (cf. https://docs.kernel.org/admin-guide/abi-testing.html#abi-file-testing-sysfs-bus-rbd)

### CODE REVIEW

**Comment by olivercalder** [comment]
Thanks for adjusting the PR title! It seems that the commit message must also match the same pattern. Would you mind please amending the commit message and force-pushing?

**Comment by sabaini** [comment]
Thanks @olivercalder for the feedback, I've pushed some updates. I see there's some test faults but not sure they are related?

**Comment by olivercalder** [comment]
Thanks! Yes, the test failures do seem unrelated. It's possible a rebase on master would solve this, but we can force-merge if necessary --- we generally prefer to avoid force pushes once the review process has begun (aside from commit messages, for example), until we're ready to merge. And we can always squash merge, so we prefer being able to follow the commit history during the review.

Out of curiosity, how is this interface intended to relate to the existing microceph interface? I see that the latter only allows access to the microceph socket, while this allows access to many system resources. What kinds of snaps are expected to use one or the other (or both?).

Since this interface grants rather substantial permissions, I'm going to request approval from @pedronis before merging. He's out until mid-next week, so I'll ask @alfonsosanchezbeato to take a look as well in the meantime.

File: interfaces/builtin/microceph_support.go
Comment (pedronis) [inline interfaces/builtin/microceph_support.go:29]: this needs to define and use also a  microcephSupportBaseDeclarationSlots similar to what microstack-support does

**Review by pedronis — DISMISSED** [review]
thanks, looking good but needs policy also for the slot side

**Review by pedronis — APPROVED** [review]
thank you

**Review by alexmurray — COMMENTED** [review]
I wonder if instead of another special-purpose interface, whether the existing block-devices and the newly proposed bcache interface would cover most of the uses-cases here - https://github.com/snapcore/snapd/pull/13523? I don't think we have anything for the rbd devices yet but perhaps we could extend block-devices to cover this as well? Thoughts? cc @pedronis

**Comment by sabaini** [comment]
> I wonder if instead of another special-purpose interface, whether the existing block-devices and the newly proposed bcache interface would cover most of the uses-cases here - #13523? I don't think we have anything for the rbd devices yet but perhaps we could extend block-devices to cover this as well? Thoughts? cc @pedronis

@alexmurray @pedronis apart from rbd devices and bcache, block-devices also lacks access to individual partitions and device mapper block devs -- we would have to add those to block-devices or introduce an extra interface for these. Would this be possible?

**Comment by alexmurray** [comment]
Apologies - I thought block-devices did include the individual partitions as well as the underlying devices - I suspect there was a Good Reason™ why individual partitions were not included in block-devices but I can't say I know what it was - @pedronis can you recall?

Perhaps for now it is better then to have this special purpose interface if there is no other way to access the individual partitions.

File: interfaces/builtin/microceph_support_test.go
Comment (Meulengracht) [inline interfaces/builtin/microceph_support_test.go:76]: This will fail now after some changes to master, please rebase on top of master and change this to whats now done in other interfaces

**Comment by sabaini** [comment]
Hey @Meulengracht , I've rebased and updated the unit test, hope this is in line with what you requested. Unit tests should pass, I can see some other CI faults but not sure if those are related?

**Comment by Meulengracht** [comment]
> Hey @Meulengracht , I've rebased and updated the unit test, hope this is in line with what you requested. Unit tests should pass, I can see some other CI faults but not sure if those are related?

They are not related, sorry for the slow answer. I'm rerunning some of the failed tests.

**Comment by pedronis** [comment]
@alexmurray as discussed separately I would be fine with adding access to partitions to `block-devices` controlled via an attribute to not change the expectations related to the interface as is too much

**Review by alexmurray — APPROVED** [review]
For the purpose of the needs of microceph, LGTM, however ideally the block-devices interface would be updated to include access to device partitions (via an interface attribute, as suggested by @pedronis) and the rbd paths would also be added to block-devices, which would then allow this interface to be deprecated in the future and microceph could then just use block-devices directly without this special-purpose interface.

**Comment by olivercalder** [comment]
For the sake of time, I'm going to rebase this on master and fix the merge conflict with `interfaces/policy/basedeclaration_test.go`, then force-push to this branch. @sabaini if you disagree, please force push your local branch to override my changes.

**Comment by olivercalder** [comment]
Actually, @sabaini I've preserved a copy of your branch at https://github.com/olivercalder/snapd/tree/microceph-support-interface

**Comment by Meulengracht** [comment]
There are consistent spread failures in this PR that needs to be handled, I can't merge this until they are fixed

**Comment by olivercalder** [comment]
Rebased again to pull in test fixes.

**Pull request merged**
